### PR TITLE
Generic worker

### LINKED ISSUE #737: Use a generic worker
All the workers use the same codebase, but we assign only one processing step to each worker at startup. We could instead allow a worker to handle all (or part of) the processing steps and let the queue manager give it the most adequate job.

By the way, the current way of doing is a particular case, where the set of supported processing steps for each worker only contain one element.

In the transition, we could have part of the workers working as of now, while some other workers are "generic", or manage a set of several steps...

The benefit would be to make it a lot easier to handle the deployment and scaling of workers, and avoiding lazy workers.

### PR DESCRIPTION
**Pull request by severo**
This PR allows a worker to process different job types. We can still dedicate workers to a sublist of jobs using a comma-separated list of the jobs in `WORKER_LOOP_ONLY_JOB_TYPES`.

This way, we will be able to reduce the allocated but unused resources.

As you can see in chart/env/prod.yaml, I let all the previous workers but added 20 instances of "genericWorker" that can process any job. If it works well, we can reduce the number of specific workers (maybe just /first-rows and /parquet-and-dataset-info) and increase the number of generic ones.

Fixes #737.

### CODE REVIEW

File: libs/libcommon/src/libcommon/queue.py [hunk 1]
```diff
@@ -430,25 +497,26 @@ def is_job_in_process(self, dataset: str, config: Optional[str] = None, split: O
             > 0
         )
 
-    def cancel_started_jobs(self) -> None:
+    def cancel_started_jobs(self, job_type: str) -> None:
         """Cancel all started jobs."""
```
Comment (AndreaFrancis) [inline libs/libcommon/src/libcommon/queue.py:501]:
```suggestion
        """Cancel all started jobs for a given type."""
```

File: libs/libcommon/tests/test_queue.py [hunk 2]
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
...
-    queue = Queue(test_type)
+    queue = Queue()
     # add a job
-    queue._add_job(dataset=test_dataset, force=True)
+    queue._add_job(job_type=test_type, dataset=test_dataset, force=True)
     # a second call adds a second waiting job
-    queue._add_job(dataset=test_dataset)
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    queue._add_job(job_type=test_type, dataset=test_dataset)
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:41]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 3]
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
...
     # get and start the first job
     job_info = queue.start_job()
     assert job_info["type"] == test_type
     assert job_info["dataset"] == test_dataset
     assert job_info["config"] is None
     assert job_info["split"] is None
     assert job_info["force"] is True
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:49]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 4]
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
...
-    queue._add_job(dataset=test_dataset, force=True)
+    queue._add_job(job_type=test_type, dataset=test_dataset, force=True)
     with pytest.raises(EmptyQueueError):
         # but: it's not possible to start two jobs with the same arguments
         queue.start_job()
     # finish the first job
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is not empty
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:59]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 5]
```diff
@@ -68,7 +70,7 @@ def test__add_job() -> None:
     # finish it
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is empty
-    assert queue.is_job_in_process(dataset=test_dataset) is False
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is False
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:73]:
```suggestion
    assert not queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 6]
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
...
-    queue = Queue(test_type)
+    queue = Queue()
     # upsert a job
-    queue.upsert_job(dataset=test_dataset, force=True)
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, force=True)
     # a second call creates a second waiting job, and the first one is cancelled
-    queue.upsert_job(dataset=test_dataset)
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    queue.upsert_job(job_type=test_type, dataset=test_dataset)
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:88]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 7]
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
...
     # the queue is not empty
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
     # process the second job
     job_info = queue.start_job()
     assert job_info["force"] is False  # the new jobs does not inherit from started forced jobs
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is empty
-    assert queue.is_job_in_process(dataset=test_dataset) is False
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is False
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:111]:
```suggestion
    assert not queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 8]
```diff
@@ -152,16 +156,20 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
...
     test_dataset = "test_dataset"
     test_config = "test_config"
-    queue = Queue(test_type, max_jobs_per_namespace=max_jobs_per_namespace)
-    queue.upsert_job(dataset=test_dataset, config=test_config, split="split1")
-    assert queue.is_job_in_process(dataset=test_dataset, config=test_config, split="split1") is True
-    queue.upsert_job(dataset=test_dataset, config=test_config, split="split2")
-    queue.upsert_job(dataset=test_dataset, config=test_config, split="split3")
+    queue = Queue(max_jobs_per_namespace=max_jobs_per_namespace)
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
+    assert (
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:161]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
```

File: libs/libcommon/tests/test_queue.py [hunk 9]
```diff
@@ -152,16 +156,20 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
...
+    )
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, config=test_config, split="split2")
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, config=test_config, split="split3")
     job_info = queue.start_job()
     assert job_info["dataset"] == test_dataset
     assert job_info["config"] == test_config
     assert job_info["split"] == "split1"
-    assert queue.is_job_in_process(dataset=test_dataset, config=test_config, split="split1") is True
+    assert (
+        queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1") is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:171]:
```suggestion
        queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
```

File: libs/libcommon/tests/test_queue.py [hunk 10]
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
...
+        ("test_type", ["test_type"]),
+        ("test_type", ["other_type", "test_type"]),
+        ("test_type", ["other_type"]),
+    ],
+)
+def test_only_job_types(job_type: str, only_job_types: Optional[list[str]]) -> None:
+    test_dataset = "test_dataset"
+    queue = Queue(max_jobs_per_namespace=100)
+    queue.upsert_job(job_type=job_type, dataset=test_dataset, config=None, split=None)
+    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:204]:
```suggestion
    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None)
```

File: libs/libcommon/tests/test_queue.py [hunk 11]
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
...
+    if only_job_types is not None and job_type not in only_job_types:
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:205]:
```suggestion
    if only_job_types and job_type not in only_job_types:
```

File: libs/libcommon/tests/test_queue.py [hunk 12]
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
...
+        with pytest.raises(EmptyQueueError):
+            job_info = queue.start_job(only_job_types=only_job_types)
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:207]: Do you use job_info?

File: libs/libcommon/tests/test_queue.py [hunk 13]
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
...
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
     # get and start the last job
     job_info = queue.start_job()
     assert job_info["type"] == test_type
     assert job_info["dataset"] == test_dataset
     assert job_info["config"] is None
     assert job_info["split"] is None
     assert job_info["force"] is True  # the new job inherits from waiting forced jobs
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:96]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: libs/libcommon/tests/test_queue.py [hunk 14]
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
...
-    queue.upsert_job(dataset=test_dataset, force=False)
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, force=False)
     with pytest.raises(EmptyQueueError):
         # but: it's not possible to start two jobs with the same arguments
         queue.start_job()
     # finish the first job
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is not empty
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
Comment (AndreaFrancis) [inline libs/libcommon/tests/test_queue.py:105]:
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

File: services/worker/tests/test_loop.py [hunk 15]
```diff
@@ -52,18 +51,21 @@ def test_process_next_job(
...
+        max_jobs_per_namespace=app_config.queue.max_jobs_per_namespace,
     )
     assert loop.process_next_job() is False
     dataset = "dataset"
     config = "config"
     split = "split"
-    loop.queue.upsert_job(dataset=dataset, config=config, split=split)
-    loop.queue.is_job_in_process(dataset=dataset, config=config, split=split) is True
+    loop.queue.upsert_job(job_type=test_processing_step.job_type, dataset=dataset, config=config, split=split)
+    loop.queue.is_job_in_process(
```
Comment (AndreaFrancis) [inline services/worker/tests/test_loop.py:65]:
```suggestion
    assert loop.queue.is_job_in_process(
```

File: services/worker/tests/test_loop.py [hunk 16]
```diff
@@ -52,18 +51,21 @@ def test_process_next_job(
...
+        job_type=test_processing_step.job_type, dataset=dataset, config=config, split=split
+    ) is True
     assert loop.process_next_job() is True
-    loop.queue.is_job_in_process(dataset=dataset, config=config, split=split) is False
+    loop.queue.is_job_in_process(
```
Comment (AndreaFrancis) [inline services/worker/tests/test_loop.py:69]:
```suggestion
    assert loop.queue.is_job_in_process(
```

**Review by AndreaFrancis — CHANGES_REQUESTED** [review]

File: libs/libcommon/tests/test_queue.py (hunk 12 above)
Comment (severo) [reply libs/libcommon/tests/test_queue.py:201]: removing, thanks

**Comment by severo** [comment]
Thanks! I fixed based on your comments (and even removed all occurrences of `is False` or `is True`). I rebased on main after https://github.com/huggingface/datasets-server/pull/805, so there should be new changes.

**Review by AndreaFrancis — APPROVED** [review]

**Pull request merged**